Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +26,63 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmpHoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (current_ < 0)
+       return true;
+    else
+        return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime time(0,0,0,0);
+    int ms_remaining;
+      if (isCharging())
+      {
+        ms_remaining = totalAmpHoursUsed() / averageCurrent_ * HOURS_TO_MS_CONVERSION;
+      }
+
+      else
+      {
+        ms_remaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / averageCurrent_ * HOURS_TO_MS_CONVERSION;
+      }
+
+    QTime timeWhenChargedOrDepleted = time.addMSecs(ms_remaining);
+
+    return timeWhenChargedOrDepleted;
+
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
+    counter_++;
+    int deltaMseconds;
+    double chargeAh;
+
+    if (previousMilliseconds_.isNull())
+    {
+        previousMilliseconds_ = batteryData.time;
+    }
+
+    QTime presentTime = batteryData.time;
+
+    deltaMseconds = abs(presentTime.msecsTo(previousMilliseconds_));
+
+    double deltaHours = deltaMseconds * MS_TO_HOURS_CONVERSION;
+
+    previousMilliseconds_ = batteryData.time;
+
+    current_ = batteryData.current;
+
+    totalCurrent_ = batteryData.current + totalCurrent_;
+
+    averageCurrent_ = totalCurrent_ / counter_;
+
+    chargeAh =  averageCurrent_ * deltaHours;
+
+    totalAmpHoursUsed_ = initialAhUsed_ + chargeAh;
```

Comment:
Don't need this many extra lines.

Group blocks of logic together.

Also `totalAmpHoursUsed_ = initialAhUsed_ + chargeAh;` has a bug